docs/api_specs/includes/watcher_api_specs.md

```diff
@@ -0,0 +1,691 @@
+        "eth_height"  : 3521678,
+        "utxo_pos"  : 10001001,
+        "owner"  : "B3256026863EB6AE5B06FA396AB09069784EA8EA",
+        "currency"  : "0000000000000000000000000000000000000000",
+        "amount" : 100
+    }
+}
+```
+
+Indicates that an invalid exit is dangerously close to finalization and hasn't been challenged. User should exit.
+
+
+#### `invalid_block`
+> An invalid_block event
+
+```json
+{
+    "event": "invalid_block",
+    "details": {
+        "blockhash"  : "DB32876CC6F26E96B9291682F3AF4A04C2AA2269747839F14F1A8C529CF90225",
+        "blocknum"  : 10000,
+    }
+}
+```
+
+A block containing an invalid tx has been processed [are there other reasons a block can be invalid?]. User should exit.
+
+
+#### `block_withholding`
+> A block_withholding event
+
+```json
+{
+    "event": "block_withholding",
+    "details": {
+        "blockhash"  : "DB32876CC6F26E96B9291682F3AF4A04C2AA2269747839F14F1A8C529CF90225",
+        "blocknum"  : 10000,
+    }
+}
+```
+
+The ChildChain is withholding a block whose hash has been published on the root chain. User should exit.
+
+#### `noncanonical_ife`
+> A noncanonical_ife event
+
+```json
+{
+    "event": "noncanonical_ife",
+    "details": {
+        "inflight_blocknum"  : 10000,
+        "inflight_tx_index"  : 12,
+        "inflight_input_index"  : 1,
+        "competing_blocknum"  : 10000,
+        "competing_tx_index"  : 10,
+        "competing_input_index"  : 0,
+    }
+}
+```
+
+An non-canonical in-flight exit has been started. It should be challenged. 
+<aside class="warning"> Not Implemented Yet.</aside> 
+
+Event details:
+Attribute | Type | Description
+--------- | ------- | -----------
+inflight_blocknum | integer | Block number of the in-flight transaction
+inflight_tx_index | integer | Index of the in-flight transaction
+inflight_input_index | integer | Index of the double-spent input in the in-flight transaction
+competing_blocknum | integer | Block number of the spending transaction
+competing_tx_index | integer | Index of the spending transaction
+competing_input_index | integer | Index of the double-spent input in the competing transaction
+
+#### `invalid_ife_challenge`
+> A invalid_ife_challenge event
+
+```json
+{
+    "event": "invalid_ife_challenge",
+    "details": {
+        "inflight_blocknum"  : 10000,
+        "inflight_tx_index"  : 12,
+    }
+}
+```
+
+A canonical in-flight exit has been challenged. The challenge should be responded to. 
+<aside class="warning"> Not Implemented Yet.</aside> 
+
+Event details:
+Attribute | Type | Description
+--------- | ------- | -----------
+inflight_blocknum | integer | Block number of the in-flight transaction
+inflight_tx_index | integer | Index of the in-flight transaction
+
+#### `piggyback_available`
+> A piggyback_available event
+
+```json
+{
+    "event": "piggyback_available",
+    "details": {
+        "inflight_blocknum"  : 10000,
+        "inflight_tx_index"  : 12,
+        "inflight_io_index"  : 1
+    }
+}
```

Comment:
Let's assume there is `InFlight transaction` with 4 inputs and 4 outputs then we're going to emit 8 events, right? Maybe it will be better to emit one `piggyback_available` event per `InFlifht transacion`

Reply:
We must design this event together with `inflight_exit.get_piggyback_data` endpoint to be elegant from a user flow perspective.

Reply:
Ok, makes sense. 
(they're not going to be 'events' anymore, they'll be entries in a list)

So, if I understand correctly:
- there's an entry for every in-flight exit with just {blocknum, txindex}
- the user needs to check if they own any inputs or outputs in the tx
- if they do, they need to get the piggyback data for the tx
- then they need to call RootChain.piggybackInFlightExit() for each input or output that they own

Reply:
Exactly like that :+1: